src/sgame/sg_bot_skilltree.cpp

```diff
@@ -0,0 +1,208 @@
+
+	// humans
+	{ "predict-aim",        BOT_H_PREDICTIVE_AIM, 5, pred_human, {} },
+};
+
+static const std::vector<botSkillTreeElement_t> initial_unlockable_skills = {
+	// movement skills
+	{ "movement", BOT_B_BASIC_MOVEMENT, 2, pred_always, movement_skills },
+	// fighting skills
+	{ "fighting", BOT_B_BASIC_FIGHT, 3, pred_always, fighting_skills },
+	// situation awareness and survival
+	{ "feels-pain", BOT_B_PAIN, 2, pred_always, survival_skills },
+};
+
+// Note: this function modifies possible_choices to remove the one we just
+//       chose, and add the new unlocked skills to that list.
+static Util::optional<botSkillTreeElement_t> ChooseOneSkill(const gentity_t *bot, skillSet_t skillSet, std::vector<botSkillTreeElement_t> &possible_choices, std::mt19937_64& rng)
+{
+	int total_skill_points = 0;
+	for (const auto &skill : possible_choices)
+	{
+		if (skill.predicate(bot, skillSet))
+		{
+			total_skill_points += skill.cost;
+		}
+	}
+
+	if (total_skill_points <= 0)
+		return Util::nullopt;
+
+	std::uniform_int_distribution<int> dist(0, total_skill_points - 1);
+	int random = dist(rng);
+
+	// find which skill this corresponds to (higher point skills are more
+	// likely to be found)
+	int cursor = 0;
+	for (auto skill = possible_choices.begin(); skill != possible_choices.end(); skill++)
+	{
+		if (!skill->predicate(bot, skillSet))
+		{
+			continue;
+		}
+
+		cursor += skill->cost;
+		if (random < cursor)
+		{
+			botSkillTreeElement_t result = *skill;
+
+			// update the list of future choices
+			possible_choices.erase(skill);
+			for (const auto &child : result.unlocked_skills)
+			{
+				possible_choices.push_back(std::move(child));
+			}
+			result.unlocked_skills.clear();
+
+			return result;
+		}
+	}
+	return Util::nullopt;
+}
+
+// This selects a set of skills for a bot.
+//
+// The algorithm is stable and deterministic.
+// Deterministic:
+//     a bot's skill set is determined only by the bot's name and the bot's
+//     skill level. A different bot will have a different skillset, but this
+//     means that a bot with the same name will have the same skillset from
+//     a game to another.
+// Stable:
+//     when a bot changes skill level, the first skills unlocked are common
+//     between both skill levels. This means that a bot whose skill is raised
+//     by BotSetSkillLevel will keep the skills it already had, and get new
+//     ones.
+//     This essentially means that a bot will keep its "personality" after
+//     a skill change, although keeping its base skills or unlocking new ones.
+std::pair<std::string, skillSet_t> BotDetermineSkills(gentity_t *bot, int skill)
+{
+	std::vector<botSkillTreeElement_t> possible_choices = initial_unlockable_skills;
+	// aliens have 51 points to spend max
+	// humans have 40 points to spend max
+	// here we give a bit more money to humans because they have more
+	// expensive skills
+	float max = G_Team(bot) == TEAM_ALIENS ? 51.0f : 46.0f;
+
+	// unlock every skill at skill 8
```

Comment:
If this is part of the design then it should be calculated by summing up the skill costs rather than counting on people to change these numbers.

Reply:
I don't want to set that into stone because 1. I'd like to allow tuning this, for example raising the human max skill count after I add a second aim skill or a precise goon barb skill and 2. because it's impossible to have a reliable count for arbitrary predicate anyway, for example if you have one skill hidden behind another, but outside of the tree.

Reply:
OK you can merge it as-is for now, but we should open an issue about making the skill tree maintainable. It's really hard to understand what I should do to these numbers if I add a new skill.

Reply:
Fair enough, I would even mostly agree on that. This is a problem with most things gameplay related though — it's hard to find a balance.